source/extensions/filters/network/redis_proxy/command_splitter_impl.cc

```diff
@@ -900,6 +901,293 @@ void MSETRequest::onChildResponse(Common::Redis::RespValuePtr&& value, uint32_t
+      std::string indexStr = std::to_string(shard_index);
+      cursor += std::string(4 - indexStr.length(), '0') + indexStr;
+    }
+    if (cursor == "0" && shard_index+1 < num_of_Shards_) {
+      std::string nextIndexStr = std::to_string(shard_index+1);
+      cursor += std::string(4 - nextIndexStr.length(), '0') + nextIndexStr;
+    }
+
+    // Response array will be created in the following format
+    // [0] -> latest scan cursor for next iteration
+    // [1] -> array for returned objects from multiple pending responses
+
+    // Setting the cursor from last response, since that is the latest one
+    // We need iterate from here in the next request.
+    Common::Redis::RespValue cstr;
+    cstr.type(Common::Redis::RespType::BulkString);
+    cstr.asString() = std::move(cursor);
+    response->asArray().emplace_back(std::move(cstr));
+
+    // Creating a temporary array to hold the objects from multiple pending responses
+    Common::Redis::RespValue objArray;
+    objArray.type(Common::Redis::RespType::Array);
+
+    // Iterate through the pending responses
+    for (size_t i = 0; i < pending_responses_.size(); ++i) {
+      if (pending_responses_[i] != nullptr) {
+        auto& resp = pending_responses_[i];
+        if (resp->type() == Common::Redis::RespType::Array) {
+            auto& obj = resp->asArray()[1];
+            if (obj.type() == Common::Redis::RespType::Array) {
+              // Iterate through the inner objects array and add its elements to the object array
+              for (size_t k = 0; k < obj.asArray().size(); ++k) {
+                objArray.asArray().emplace_back(std::move(obj.asArray()[k]));
+              }
+            }
+        }
+      }
+    }
+    pending_responses_.clear();
+    // Add the object array to the main response array as a nested array
+    response->asArray().emplace_back(std::move(objArray));
+
+    if (response->type() == Common::Redis::RespType::Error){
+      ENVOY_LOG(debug,"recived error for index: '{}'", shard_index);
+      Common::Redis::RespValuePtr response_t = std::move(response);
+      ENVOY_LOG(debug, "response: {}", response_t->toString());
+      callbacks_.onResponse(std::move(response_t));
+    } else {
+      updateStats(error_count_ == 0);
+      Common::Redis::RespValuePtr response_t = std::move(response);
+      ENVOY_LOG(debug, "response: {}", response_t->toString()); 
+      callbacks_.onResponse(std::move(response_t));
+    }
+  }
+}
+
+void ScanRequestBase::onChildFailure(int32_t reqindex,int32_t shardindex) {
+  updateStats(false);
+  onChildResponse(Common::Redis::Utility::makeError(Response::get().UpstreamFailure), reqindex,shardindex);
+}
+
+ScanRequestBase::~ScanRequestBase() {
+#ifndef NDEBUG
+  for (const PendingRequest& request : pending_requests_) {
+    ASSERT(!request.handle_);
+  }
+#endif
+}
+
+void ScanRequestBase::cancel() {
+  for (PendingRequest& request : pending_requests_) {
+    if (request.handle_) {
+      request.handle_->cancel();
```

Comment:
no need to send cancel for all the PendingRequest& request : pending_requests_ , send it only for request in pipeline

Reply:
We are sending cancel only to the requests having handle_ and all the pending requests will be cleared when the unique pointer created for the scan request goes out of scope. It goes out of scope in the proxy filter.